lte/gateway/python/magma/pipelined/ebpf/ebpf_manager.py

```diff
@@ -0,0 +1,222 @@
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import ctypes
+import logging
+import socket
+import struct
+import subprocess
+from builtins import input
+from socket import AF_INET, htons
+from subprocess import call
+from sys import argv
+from threading import Thread
+
+import netifaces
+from bcc import BPF
+from magma.pipelined.mobilityd_client import get_mobilityd_gw_info
+from pyroute2 import IPDB, IPRoute, NetlinkError, NetNS, NSPopen
+from scapy.layers.l2 import getmacbyip
+
+LOG = logging.getLogger("pipelined.ebpf")
+# LOG.setLevel(logging.DEBUG)
+
+BASE_MAP_FS = "/sys/fs/bpf/"
+BPF_UL_FILE = "/var/opt/magma/ebpf/ebpf_ul_handler.c"
+UL_MAP_NAME = "ul_map"
+
+bpf_man = None
+
+"""
+    Use pipelineD configuration to initialize eBPF manager for AGW.
+"""
+
+
+def get_ebpf_manager(config):
+    global bpf_man
+    if bpf_man:
+        return bpf_man
+
+    if 'ebpf' in config:
+        enabled = config['ebpf']['enabled']
+    else:
+        enabled = False
+    gw_info = get_mobilityd_gw_info()
+    for gw in gw_info:
+        if gw.vlan == "":
+            bpf_man = bpf_manager(config['nat_iface'], config['enodeb_iface'], gw.ip, enabled)
+            if enabled:
+                # TODO: For Development purpose dettch and attach latest eBPF code.
+                # Remove this for production deployment
+                bpf_man.detach_ul_ebpf()
+                bpf_man.attach_ul_ebpf()
+                LOG.info("eBPF manager: initilized: enabled: %s", enabled)
+            return bpf_man
+
+    LOG.info("eBPF manager: Not initilized")
+    return None
+
+
+"""
+    eBPF manager for AGW.
+    Initialize eBPF based datapath for AGW as per the pipelineD config.
+
+    Returns:
+        eBPF manager object.
+"""
+
+
+class bpf_manager:
+    def __init__(self, sgi_if_name: str, s1_if_name: str, gw_ip: str, enabled=True, bpf_file: str = BPF_UL_FILE):
+
+        self.b = BPF(src_file=bpf_file, cflags=[''])
+        self.s1_fn = self.b.load_func("gtpu_get_tunnel", BPF.SCHED_CLS)
+        self.ul_map = self.b.get_table(UL_MAP_NAME)
+        self.sgi_if_name = sgi_if_name
+        self.s1_if_name = s1_if_name
+        self.ul_src_mac = self._get_mac_address(sgi_if_name)
+        self.ul_gw_mac = self._get_mac_address_of_ip(gw_ip)
+        self.sgi_if_index = self._get_ifindex(self.sgi_if_name)
+        self.enabled = enabled
+
+    """Attach eBPF Uplink traffic handler
+    """
+
+    def attach_ul_ebpf(self):
+        s1_if_index = self._get_ifindex(self.s1_if_name)
+
+        ipr = IPRoute()
+        try:
+            ipr.tc("add", "ingress", s1_if_index, "ffff:")
```

Comment:
we can also explore this type of filter
```
    ipr.tc("add", "clsact", ifindex)
    action = {"kind": "bpf", "fd": fn.fd, "name": fn.name, "action": "ok"}
    ipr.tc(
        "add-filter", "u32", ifindex, ":1", parent="ffff:fff3", action=[action],
        protocol=protocols.ETH_P_ALL, classid=1, target=0x10002, keys=['0x0/0x0+0'],
    )
```
Also adding action that way allow to control the packet path directly, we would be able to do `return bpf_redirect(if_idx,0);` instead of doing bpf clone redirect

Reply:
switched to clsact.